src/Grpc.AspNetCore.Web/Internal/Base64PipeReader.cs

```diff
@@ -0,0 +1,193 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Buffers;
+using System.Buffers.Text;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Grpc.AspNetCore.Web.Internal
+{
+    /// <summary>
+    /// Reads and decodes base64 encoded bytes from the inner reader.
+    /// </summary>
+    internal class Base64PipeReader : PipeReader
+    {
+        private readonly PipeReader _inner;
+        private ReadOnlySequence<byte> _currentInnerBuffer;
+        private ReadOnlySequence<byte> _currentDecodedBuffer;
+
+        public Base64PipeReader(PipeReader inner)
+        {
+            _inner = inner;
+        }
+
+        public override void AdvanceTo(SequencePosition consumed)
+        {
+            var consumedPosition = ResolvePosition(consumed);
+
+            _inner.AdvanceTo(consumedPosition);
+        }
+
+        public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
+        {
+            var consumedPosition = ResolvePosition(consumed);
+            var examinedPosition = ResolvePosition(examined);
+
+            _inner.AdvanceTo(consumedPosition, examinedPosition);
+        }
+
+        private SequencePosition ResolvePosition(SequencePosition base64Position)
+        {
+            // TODO - is there a better way to get the index of a position in a sequence?
+            var length = _currentDecodedBuffer.Slice(0, base64Position).Length;
```

Comment:
I think this is the standard way to get the length between SequencePositions, but I agree it would be nice if there was a helper method to avoid the intermediate ReadOnlySequence. The good news is that neither Slice() nor the Length property loop over the individual segments\*. ReadOnlySequence.Length uses the segment's "RunningIndex" to make calculating the length O(1).

https://github.com/dotnet/runtime/blob/377b3c7c3e364b3183696744c6a5c6968bcdffd4/src/libraries/System.Memory/src/System/Buffers/ReadOnlySequence.Helpers.cs#L501-L518

\* Slice will loop over segments if given a longs/ints instead of SequencePositions. But in this case, since the only long/int being passed in is 0 as the start index, looping should not happen in practice.